### PR TITLE
Allow search jobs to be created if going into a queued state

### PR DESCRIPTION
**Pull request by shakeelmohamed**

### CODE REVIEW

File: src/Splunk.Client/Splunk/Client/JobCollection.cs [hunk 1]
```diff
@@ -234,12 +259,27 @@ async Task<Job> CreateAsync(IEnumerable<Argument> arguments, DispatchState requi
 
             Job job = new Job(this.Context, this.Namespace, name: searchId);
 
-            await job.GetAsync().ConfigureAwait(false);
+            await job.GetAsync(requiredState).ConfigureAwait(false);
```
Comment (shakeelmohamed) [inline src/Splunk.Client/Splunk/Client/JobCollection.cs]: Previously in `CreateAsync()` we would ignore the `requiredState` value and by default wait for every job to reach a created state before returning - resulting in timeouts.

File: src/Splunk.Client/Splunk/Client/DateTimeConverter.cs [hunk 2]
```diff
@@ -58,9 +58,14 @@ public override DateTime Convert(object input)
                 return x.Value;
             }
 
-            DateTime value;
+            var dateString = input.ToString();
+            if (dateString.Trim().Equals(""))
+            {
+                return DateTime.MinValue;
```
Comment (shakeelmohamed) [inline src/Splunk.Client/Splunk/Client/DateTimeConverter.cs]: The original XML parsing bug was addressed with this change

File: src/Splunk.Client/Splunk/Client/JobCollection.cs [hunk 3]
```diff
@@ -195,6 +195,31 @@ public override async Task<Job> CreateAsync(IEnumerable<Argument> arguments)
             return job;
         }
 
+        public virtual async Task<Job> Create(string search, int count = 0, // This version doesn't wait
```
Comment (shakeelmohamed) [inline src/Splunk.Client/Splunk/Client/JobCollection.cs]: We can remove these, but I wanted to add another `Create()` function that doesn't wait until a `requiredStatus` is reached

File: src/Splunk.Client/Splunk/Client/JobCollection.cs [hunk 4]
```diff
@@ -234,12 +259,27 @@ async Task<Job> CreateAsync(IEnumerable<Argument> arguments, DispatchState requi
...
             Job job = new Job(this.Context, this.Namespace, name: searchId);
 
-            await job.GetAsync().ConfigureAwait(false);
+            await job.GetAsync(requiredState).ConfigureAwait(false);
             await job.TransitionAsync(requiredState).ConfigureAwait(false);
 
             return job;
         }
 
+       async Task<Job> Create(IEnumerable<Argument> arguments) // this version doesn't wait
```
Comment (shakeelmohamed) [inline src/Splunk.Client/Splunk/Client/JobCollection.cs]: internal helper used by the new `Create()` above

File: test/acceptance-tests/TestSearchJob.cs [hunk 5]
```diff
@@ -485,6 +485,91 @@ public async Task CanRefreshJob()
...
+                List<Job> jobs = new List<Job>();
+                Job job = null;
+                do
+                {
+                    JobArgs jobArgs = new JobArgs();
+                    jobArgs.SearchMode = SearchMode.RealTime;
+
+                    try
+                    {
+                        job = await service.Jobs.Create(searchPrefix + i.ToString(), args: jobArgs);
```
Comment (shakeelmohamed) [inline test/acceptance-tests/TestSearchJob.cs]: This test uses the new `Create()`

File: src/Splunk.Client.Helpers/Splunk/Client/Helpers/SdkHelper.cs [hunk 6]
```diff
@@ -39,7 +39,7 @@ static SdkHelper()
             //// 2. Set its ServerCertificateValidationCallback
             //// 3. Instantiate a Splunk.Client.Context with the WebRequestHandler
 
-            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls;
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
```
Comment (zenmoto) [inline src/Splunk.Client.Helpers/Splunk/Client/Helpers/SdkHelper.cs]: Does moving from Tls to Tls12 have any downsides?

File: test/acceptance-tests/TestSearchJob.cs [hunk 7]
```diff
@@ -485,6 +485,91 @@ public async Task CanRefreshJob()
...
+                Job job = null;
+                do
+                {
+                    JobArgs jobArgs = new JobArgs();
+                    jobArgs.SearchMode = SearchMode.RealTime;
+
+                    try
+                    {
+                        // Allow for jobs to be queued w/o waiting for them to get to running state
+                        job = await service.Jobs.CreateAsync(searchPrefix + i.ToString(), args: jobArgs, requiredState: DispatchState.Queued);
```
Comment (shakeelmohamed) [inline test/acceptance-tests/TestSearchJob.cs]: Here we pass in the `requiredState` of queued so we don't keep retrying forever to reach a running state, which isn't possible

File: src/Splunk.Client.Helpers/Splunk/Client/Helpers/SdkHelper.cs (hunk 6 above)
Comment (shakeelmohamed) [reply src/Splunk.Client.Helpers/Splunk/Client/Helpers/SdkHelper.cs]: This is a test helper so no

File: src/Splunk.Client/Splunk/Client/JobCollection.cs (hunk 1 above)
Comment (zenmoto) [reply src/Splunk.Client/Splunk/Client/JobCollection.cs]: This version will still time out if required state is running and we get queued though?

File: src/Splunk.Client.Helpers/Splunk/Client/Helpers/SdkHelper.cs (hunk 6 above)
Comment (zenmoto) [reply src/Splunk.Client.Helpers/Splunk/Client/Helpers/SdkHelper.cs]: Ah.  Right.

File: test/acceptance-tests/TestSearchJob.cs (hunk 7 above)
Comment (zenmoto) [reply test/acceptance-tests/TestSearchJob.cs]: Seems like we need to be able to wait on _either_ running or queued in most cases (if we are going to wait).

File: src/Splunk.Client/Splunk/Client/Job.cs [hunk 8]
```diff
@@ -470,7 +471,8 @@ public virtual async Task GetAsync(DispatchState dispatchState, int delay = 3000
                             await response.EnsureStatusCodeAsync(HttpStatusCode.OK).ConfigureAwait(false);
                             await this.ReconstructSnapshotAsync(response).ConfigureAwait(false);
 
-                            if (this.DispatchState >= dispatchState)
+                            // None or Parsing are the only states to continue checking status for
+                            if (this.DispatchState >= dispatchState || this.DispatchState == DispatchState.Queued)
```
Comment (shakeelmohamed) [inline src/Splunk.Client/Splunk/Client/Job.cs]: @zenmoto see the change here

File: test/acceptance-tests/TestSearchJob.cs (hunk 7 above)
Comment (shakeelmohamed) [reply test/acceptance-tests/TestSearchJob.cs]: done

**Pull request merged**